app/src/main/java/de/westnordost/streetcomplete/overlays/places/PlacesOverlayForm.kt

```diff
@@ -259,24 +259,34 @@ private suspend fun createEditAction(
     val tagChanges = StringMapChangesBuilder(element?.tags ?: emptyMap())
 
     val hasAddedNames = previousNames.isEmpty() && newNames.isNotEmpty()
-    val hasChangedNames = previousNames != newNames
+    var hasChangedNames = previousNames != newNames
     val hasChangedFeature = newFeature != previousFeature
+    val newFeatureType = if (newFeature.isSuggestion) newFeature.id.substringBeforeLast("/") else newFeature.id
+    val previousFeatureType = if (previousFeature?.isSuggestion == true) previousFeature?.id?.substringBeforeLast("/") else previousFeature?.id
+    val hasChangedFeatureType = previousFeatureType != newFeatureType
     val isFeatureWithName = newFeature.addTags.get("name") != null
     val wasFeatureWithName = previousFeature?.addTags?.get("name") != null
     val wasVacant = element != null && element.isDisusedPlace()
     val isVacant = newFeature.id == "shop/vacant"
 
+    if (newFeature.isSuggestion) {
+        // selecting NSI preset will always return empty newNames, even if NSI does set new name=* tag
+        hasChangedNames = listOf(LocalizedName("", newFeature.addTags.get("name").orEmpty())) != previousNames
+    }
```

Comment:
You can call `parseLocalizedNames(newFeature.addTags)` to get a list of `LocalizedName`.

Also, what about `previousFeature`?

Reply:
> Also, what about previousFeature?

This "hack" for `parseLocalizedNames(newFeature.addTags)` is needed, because `newNames` are empty if NSI is chosen (as opposed as to when regular type+name are entered in separate fields, when `newNames` is set), so we have to fetch the new name (which would be set by preset) **ourselves** if it is NSI.

On the other hand, such `parseLocalizedNames(previousFeature.addTags)` hack is not needed because `previousNames` seems to be supplied _regardless_ of whether previous value was type+name or NSI.

(I've put some `Log.w` debugs in that [debug apk](https://github.com/mnalis/StreetComplete/actions/runs/11154983310) while I was testing, and that is what they've shown)

If that is what you meant?

Reply:
That's what I meant. Alright